File: app/src/main/java/algorithm/study/dohyun/programmers/Prg_178871.java
package algorithm.study.dohyun.programmers;
import java.util.*;
public class Prg_178871 {
    public static void main(String[] args){
        Prg_178871 sol = new Prg_178871();

        String[] players = { "mumu", "soe", "poe", "kai", "mine" };
        String[] callings = { "kai", "kai", "mine", "mine" };

        String[] ans = sol.solution(players,callings);
        for(String i:ans)
            System.out.print(i+" ");
        System.out.println();
    }
    public String[] solution( String[] players, String[] callings){

        // hashmap 이름이랑 등수 넣고 이름이 불리면 두 value값 변경
        HashMap<String, Integer> rank = new HashMap<>();
        for(int i=0; i< players.length; i++){
            rank.put(players[i], i);
        }
        for(String win : callings){
            // kai
            int nowrank = rank.get(win); // 3
            String after = players[nowrank -1 ]; // 앞에 가던 사람

            players[nowrank] = after;
            players[nowrank -1 ] = win;


            rank.put(win, nowrank -1 );
            rank.put(after , nowrank );
        }


        return players;
    }
}

File: app/src/main/java/algorithm/study/devjk_sample/programmers/Prg_178871.kt
package algorithm.study.devjk_sample.programmers

/**
 * 프로그래머스 - 달리기 경주
 *
 * @link https://school.programmers.co.kr/learn/courses/30/lessons/178871
 */
class Prg_178871 {

    fun solution(players: Array<String>, callings: Array<String>): Array<String> {
        val playerMap: MutableMap<String, Int> = setPlayerMap(players)
        val rankMap: MutableMap<Int, String> = setRankMap(players)

        for (prevPlayer in callings) {
            val currentRank = playerMap[prevPlayer]!!
            val postPlayer = rankMap[currentRank - 1]!!
            playerMap[prevPlayer] = currentRank - 1
            playerMap[postPlayer] = currentRank
            rankMap[currentRank] = postPlayer
            rankMap[currentRank - 1] = prevPlayer
        }

        return playerMap.toList().sortedBy { it.second }.map { it.first }.toTypedArray()
    }

    private fun setPlayerMap(players: Array<String>): MutableMap<String, Int> {
        val map = HashMap<String, Int>()
        for (i in players.indices) {
            val player = players[i]
            map[player] = i + 1
        }
        return map.toMutableMap()
    }

    private fun setRankMap(players: Array<String>): MutableMap<Int, String> {
        val map = HashMap<Int, String>()
        for (i in players.indices) {
            val player = players[i]
            map[i + 1] = player
        }
        return map.toMutableMap()
    }

}

fun main(args: Array<String>) {
    val sol = Prg_178871()
    val ans: Array<String> = sol.solution(
        arrayOf("mumu", "soe", "poe", "kai", "mine"),
        arrayOf("kai", "kai", "mine", "mine")
    )
    print("answer-- ")
    println(ans.contentToString())
}